Update reduction variables to SYCL 2020 spec behavior

fknorr opened the pull request:
Will fix #572 once complete.

## Resolved so far
- [x] Pass buffers instead of accessors into `sycl::reduction`
- [x] Provide implicit identities to `sycl::plus` and companions
- [x] Handle the `initialize_to_identity` property on reductions
- [x] Support reduction functions without an identity
- [x] Clarify whether `sycl::minimum` and `sycl::maximum` should have an identity -- the spec does not mention this, we should probably see what DPC++ is doing here and mirror the behavior. See also: KhronosGroup/SYCL-Docs#163
- [x] Properly handle implicit identities of `sycl::plus` and the like in the face of implicit argument type conversions
- [x] Come up with a less hacky implementation in `hiplike_kernel_launcher`
- [x] Document the implementation

## Open

None 

## Probably out of Scope
- Level Zero support (old API not implemented either)
- Evaluate whether the `initialized` predicate for the no-identity case can be handled more efficiently on the GPU (e.g. with bitmasks and subgroup reductions). Not investigated so far since the reduction of values themselves still has optimization potential by sharing code with (optimized) group reductions, e.g. using intrinsics like CUDA `__reduce_and_sync`.

fknorr committed:
New SYCL 2020 API for reductions

- No support for reductions without an identity so far
- init_with_identity not implemented yet

fknorr committed:
Reductions with runtime-provided identities; initialize_with_identity support

Currently no reductions without identities on GPU

fknorr committed:
Support reductions without identity on hiplike backends

fknorr committed:
Test reduction kernels with 0 or > 1 combine()s

illuhad commented:
This introduces a lot of complexity to an already complex part of the code. Can you please provide some more documentation on how it works?

Also, can we please keep the old accessor-based API for compatibility reasons? There is still a lot of DPC++ code out there that uses the original oneAPI reduction proposal, and we can currently support this quite conveniently by just adding an alias for the ONEAPI namespace.

fknorr committed:
Add minimum/maximum identities; more reduction test cases

fknorr committed:
Implement plus<void>, multiplies<void>, ... function objects

fknorr committed:
Re-introduce old accessor-based reduction interface

fknorr committed:
Simplify reduction implementation in hiplike_kernel_launcher

fknorr committed:
Reductions: Improve naming, add documentation

fknorr commented:
I have added some documentation and re-introduced the old accessor API including tests.

I think that Level Zero support and optimization beyond the original state of reductions should happen in a separate PR to keep this patch small.

fknorr committed:
Add missing reducer::identity

fknorr committed:
Add tests on sycl::known_identity

fknorr committed:
Allow cv-qualified bool in function objects for consistency

Also silence warnings about bitwise bool operators (these are verified to be correct by unit tests).

fknorr committed:
Reductions: fix indentation

fknorr commented:
@illuhad This is now complete AFAIC.

illuhad commented:
@hipSYCL-ci-bot test this please

fknorr commented:
Rebased.

fknorr commented:
Rebased to resolve conflicts; I believe the CI failure is spurious (network outage).

DieGoldeneEnte commented:
@fknorr could you create a PR for the has/get_known_identiy code (both `functional` files and the test/CMakeLists.txt)?
It would be very helpful to merge that part separately, so I can already use it in my rewrite of the group-functions.

fknorr commented:
@DieGoldeneEnte I'm reluctant to put any more work into this as it hasn't seen any action in more than a year. But you're free to extract a sub-PR of features you're interested in and submit it as your own!

fknorr commented:
Superseded by #1453 .
